src/dag.rs

```diff
@@ -599,52 +627,54 @@ impl<D: DagLike, S: SharingTracker<D>> Iterator for PostOrderIter<D, S> {
+                        current.left_idx,
+                        current.right_idx,
+                    );
+                    let already_yielded = current_index.is_some();
+                    let current_index = current_index.unwrap_or(self.index);
+
+                    // Then, update the item's parents and/or sibling, to make sure that its
+                    // parent has the correct child index. We need to do this whether or not
+                    // we're going to skip the element.
                     let stack_len = self.stack.len();
                     match current.previous {
                         Previous::Root => {
                             assert_eq!(stack_len, 0);
                         }
                         Previous::ParentLeft => {
                             assert!(self.stack[stack_len - 1].processed);
-                            self.stack[stack_len - 1].left_idx = Some(self.index);
+                            self.stack[stack_len - 1].left_idx = Some(current_index);
                         }
                         Previous::ParentRight => {
                             assert!(self.stack[stack_len - 1].processed);
-                            self.stack[stack_len - 1].right_idx = Some(self.index);
+                            self.stack[stack_len - 1].right_idx = Some(current_index);
                         }
                         Previous::SiblingLeft => {
                             assert!(self.stack[stack_len - 2].processed);
-                            self.stack[stack_len - 2].left_idx = Some(self.index);
-                            check_sibling = true;
+                            self.stack[stack_len - 2].left_idx = Some(current_index);
                         }
                     }
-                    self.tracker.record(
-                        &current.elem,
-                        self.index,
-                        current.left_idx,
-                        current.right_idx,
-                    );
-                    // There is a special case here where we are yielding a node which
-                    // is identical to its sibling (according to the sharing tracker).
-                    // In this case we have a repeated node which nonetheless made it
-                    // into the stack, so we pop it here.
-                    if check_sibling
-                        && self
-                            .tracker
-                            .seen_before(&self.stack[stack_len - 1].elem)
-                            .is_some()
-                    {
-                        self.stack[stack_len - 2].right_idx = Some(self.index);
-                        self.stack.pop();
+                    // Having updated the parent indices, so that our iterator is in a
+                    // consistent state, we can safely skip here if the current item is
+                    // already yielded.
+                    if already_yielded {
+                        continue;
```

Comment:
8e485b5ccb5130204c12fbdf5398f7acf986b3f4 We don't need to `self.stack.pop()` like before because the current item has `processed = True` and will be popped and skipped in the next loop iteration?

We should add this to the documentation.

Reply:
This question only makes sense in the context of the old (wrong) algorithm. We already popped the item at the top of the loop. There's no reason we would pop anything else here, so I don't know what I would add to the documentation.